ncm-download/src/main/perl/download.pm

```diff
@@ -1,247 +1,265 @@
-            if (!$success) {
-                $self->warn("failed to retrieve $source from any proxies, using original");
+        if ($source !~ m{^//}) {
+            if ($file->{server}) {
+                $source = "//$file->{server}$source";
+            } else {
+                # TODO: ignoring with an error?
+                #       better would be to force this being non-empty via schema
+                #       (eg no empty server attribute; or no empty server without default server specified),
+                #       and not support ignoring files with error reported
+                $self->error("$fn requested but no server, ignoring");
+                return;
             }
         }
-        if (!$success) {
-            if (!$self->download(
-                    file => $file,
-                    href => $source,
-                    timeout => $inf->{timeout},
-                    gssneg => $gss,
-                    cacert => $inf->{files}->{$f}->{cacert},
-                    capath => $inf->{files}->{$f}->{capath},
-                    cert => $inf->{files}->{$f}->{cert},
-                    key =>  $inf->{files}->{$f}->{key},
-                    min_age => $min_age)) {
-                $self->error("failed to retrieve $source, skipping");
-                next;
+
+        if ($source =~ m{^//}) {
+            # server specified, but no proto
+            if ($file->{proto}) {
+                $source = "$file->{proto}:$source";
+            } else {
+                # TODO: ignoring with error?
+                $self->error("$fn requested but no proto, ignoring");
+                return;
             }
         }
-        my @opts = ();
-        if (exists $inf->{files}->{$f}->{perm}) {
-            push(@opts, 'mode' => $inf->{files}->{$f}->{perm});
-        }
-        if (exists $inf->{files}->{$f}->{owner}) {
-            push(@opts, 'owner' => $inf->{files}->{$f}->{owner});
+    }
+
+    my %opts = (
+        href => $source,
+        gss_ccache => $file->{gss_ccache},
+        cacert => $file->{cacert},
+        capath => $file->{capath},
+        cert => $file->{cert},
+        key =>  $file->{key},
+        min_age => $file->{min_age},
+        # TODO: head_timeout used to be only per file for proxies, and per component for non-proxy
+        head_timeout => $file->{head_timeout},
+        # TODO: timeout used to be only per file for proxies, and not at all for non-proxy
+        timeout => $file->{timeout},
+        );
+
+    my $success = 0;
+    if (@$proxyhosts && $file->{proxy}) {
+        my $attempts = scalar @$proxyhosts;
+        while ($attempts--) {
+            # take the head off the list and rotate to the end...
+            # we do this in order to avoid continually trying a dead
+            # proxy - once we've found one that works, we'll keep using it.
+            my $proxy = shift @$proxyhosts;
+            $self->debug(3, "Trying proxy $proxy for source $source and filename $fn");
```

Comment:
This results in:

[DEBUG] Trying proxy ARRAY(0x3dc79c0) for source http://[redacted]/group.byname and filename /var/tmp/dsdbfiles/group.byname

which also fails.

The previous output was:
```
[DEBUG] Trying proxy [correct hostname]
[DEBUG] Processing file /var/tmp/dsdbfiles/rpc.bynumber from http://[redacted]/rpc.bynumber
[VERB] Executing command: /usr/bin/curl -s -R -f --create-dirs -o /var/tmp/dsdbfiles/rpc.bynumber http://$proxyhost/$path/rpc.bynumber
[VERB] Command options: stderr=SCALAR(0x2680ea8)
[DEBUG] We seem to have sucessfully downloaded a new copy of the file
[INFO] successfully processed file /var/tmp/dsdbfiles/rpc.bynumber
```